File: elixir/lib/fog/delta.ex
defmodule Fog.Delta do
  @moduledoc """
  The change in fog between two FogMap snapshots.

  diff/2 is a pure function. Two maps in, one delta out.

  net < 0  → fog lifted (new signal entered the system)
  net == 0 → epistemic arbitrage (ignorance redistributed, not reduced)
  net > 0  → fog deepened (agent knows less than before)
  """

  alias Fog.Map, as: FogMap

  @type t :: %__MODULE__{
    agent_id:        String.t(),
    gaps_lifted:     MapSet.t(String.t()),
    gaps_added:      MapSet.t(String.t()),
    gaps_unchanged:  MapSet.t(String.t())
  }

  defstruct [:agent_id, :gaps_lifted, :gaps_added, :gaps_unchanged]

  @doc "Compute the fog delta between two map snapshots."
  def diff(%FogMap{} = before, %FogMap{} = after) do
    before_keys = before.gaps |> Map.keys() |> MapSet.new()
    after_keys  = after.gaps  |> Map.keys() |> MapSet.new()

    %__MODULE__{
      agent_id:       before.agent_id,
      gaps_lifted:    MapSet.difference(before_keys, after_keys),
      gaps_added:     MapSet.difference(after_keys, before_keys),
      gaps_unchanged: MapSet.intersection(before_keys, after_keys)
    }
  end

  @doc "Net change in fog. Negative is good."
  def net(%__MODULE__{} = delta) do
    MapSet.size(delta.gaps_added) - MapSet.size(delta.gaps_lifted)
  end

  @doc "True if fog moved but didn't shrink — ignorance redistributed, not reduced."
  def arbitrage?(%__MODULE__{} = delta) do
    MapSet.size(delta.gaps_lifted) > 0 and net(delta) == 0
  end

  @doc "True if fog genuinely shrank — new signal entered."
  def lift?(%__MODULE__{} = delta), do: net(delta) < 0

  @doc "True if the agent knows less than before."
  def deepening?(%__MODULE__{} = delta), do: net(delta) > 0

  @doc "Human-readable summary."
  def summary(%__MODULE__{} = delta) do
    n = net(delta)
    cond do
      n == 0 and MapSet.size(delta.gaps_lifted) == 0 ->
        "[#{delta.agent_id}] static — no change"
      arbitrage?(delta) ->
        "[#{delta.agent_id}] arbitrage — " <>
        "+#{MapSet.size(delta.gaps_added)} -#{MapSet.size(delta.gaps_lifted)} net=0 " <>
        "(ignorance redistributed, not reduced)"
      lift?(delta) ->
        "[#{delta.agent_id}] lift — net=#{n} (fog clearing)"
      true ->
        "[#{delta.agent_id}] deepening — net=+#{n}"
    end
  end
end

File: haskell/src/Fog/Delta.hs
-- | FogDelta — the change between two FogMap snapshots.
--
-- diff is a pure function. Two values in, one value out.
-- The result tells you whether ignorance shrank, grew, or just moved.
module Fog.Delta
  ( FogDelta(..)
  , DeltaKind(..)
  , diff
  , deltaKind
  , netChange
  , entropyDelta
  , summarise
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Set (Set)
import qualified Data.Set as Set
import Fog.Gap (effectiveConfidence)
import Fog.Map (FogMap(..), fogKeys, fogAgentId, fogVolume, fogEntropy)

-- | What kind of change happened.
data DeltaKind
  = Static      -- ^ Nothing changed.
  | Lift        -- ^ Fog shrank — new signal entered.
  | Arbitrage   -- ^ Fog moved but didn't shrink — ignorance redistributed.
  | Deepening   -- ^ Fog grew — agent knows less than before.
  deriving (Eq, Show)

-- | The change between two fog map snapshots.
-- Stores map references for weighted computations.
data FogDelta = FogDelta
  { deltaAgentId    :: String
  , gapsLifted      :: Set String     -- ^ Present in before, absent in after.
  , gapsAdded       :: Set String     -- ^ Absent in before, present in after.
  , gapsUnchanged   :: Set String     -- ^ Present in both.
  , deltaBeforeMap  :: FogMap
  , deltaAfterMap   :: FogMap
  , deltaNow        :: Double         -- ^ POSIX time snapshot.
  , deltaHalfLife   :: Double
  } deriving (Eq, Show)

eps :: Double
eps = 1.0e-9

-- | Diff two fog maps. Pure function.
-- Pass current POSIX time and half-life for weighted net and entropy.
diff :: Double -> Double -> FogMap -> FogMap -> FogDelta
diff now hl before after = FogDelta
  { deltaAgentId   = fogAgentId before
  , gapsLifted     = Set.difference bk ak
  , gapsAdded      = Set.difference ak bk
  , gapsUnchanged  = Set.intersection bk ak
  , deltaBeforeMap = before
  , deltaAfterMap  = after
  , deltaNow       = now
  , deltaHalfLife  = hl
  }
  where
    bk = fogKeys before
    ak = fogKeys after

-- | Weighted net change in fog volume.
--
-- @
-- net = vol(added) − vol(lifted)
-- @
--
-- Uses effective_confidence as weights. Negative is good.
netChange :: FogDelta -> Double
netChange d =
  let now  = deltaNow d
      hl   = deltaHalfLife d
      eff fmap k = maybe 0.0 (effectiveConfidence now hl)
                        (Map.lookup k (fogGaps fmap))
      volAdded  = sum [ eff (deltaAfterMap  d) k | k <- Set.toList (gapsAdded  d) ]
      volLifted = sum [ eff (deltaBeforeMap d) k | k <- Set.toList (gapsLifted d) ]
  in volAdded - volLifted

-- | Change in fog entropy between before and after.
--
-- @
-- ΔH = H(after) − H(before)
-- @
--
-- ΔH < 0: fog became more concentrated — fewer, better-defined gaps.
-- ΔH > 0: fog spread out — broader, more diffuse uncertainty.
-- ΔH ≈ 0: same structural shape (even if individual gaps changed).
entropyDelta :: FogDelta -> Double
entropyDelta d =
  let now = deltaNow d
      hl  = deltaHalfLife d
  in fogEntropy now hl (deltaAfterMap d) - fogEntropy now hl (deltaBeforeMap d)

-- | Classify the delta.
deltaKind :: FogDelta -> DeltaKind
deltaKind d
  | abs n < eps, Set.null (gapsLifted d) = Static
  | n < 0                                = Lift
  | abs n < eps                          = Arbitrage
  | otherwise                            = Deepening
  where n = netChange d

fmt3 :: Double -> String
fmt3 x = show (fromIntegral (round (x * 1000) :: Int) / 1000.0 :: Double)

-- | Human-readable summary.
summarise :: FogDelta -> String
summarise d =
  let n  = netChange d
      ed = entropyDelta d
      edStr = " ΔH=" <> (if ed >= 0 then "+" else "") <> fmt3 ed
  in case deltaKind d of
    Static    ->
      "[" <> deltaAgentId d <> "] static — no change" <> edStr
    Lift      ->
      "[" <> deltaAgentId d <> "] lift — net=" <> fmt3 n <> " (fog clearing)" <> edStr
    Arbitrage ->
      "[" <> deltaAgentId d <> "] arbitrage — "
      <> "+" <> show (Set.size (gapsAdded  d))
      <> " -" <> show (Set.size (gapsLifted d))
      <> " net=" <> fmt3 n
      <> " (ignorance redistributed, not reduced)" <> edStr
    Deepening ->
      "[" <> deltaAgentId d <> "] deepening — net=+" <> fmt3 n <> edStr
